File: backend/app/schemas/apply.py
import uuid
from enum import StrEnum

from pydantic import BaseModel, Field


class ATSDifficultyEnum(StrEnum):
    easy_apply = "easy_apply"
    multi_step = "multi_step"
    manual_only = "manual_only"
    manual_assist = "manual_assist"


class ATSDetectRequest(BaseModel):
    job_id: uuid.UUID
    apply_url: str | None = None


class ATSDetectResponse(BaseModel):
    application_id: uuid.UUID
    task_id: str
    message: str


class ATSDetectionStatusResponse(BaseModel):
    application_id: uuid.UUID
    job_id: uuid.UUID
    status: str
    ats_platform: str | None = None
    ats_detection_method: str | None = None
    ats_confidence: float | None = None
    ats_form_url: str | None = None
    ats_detected_fields: list[str] | None = None
    ats_screenshot_path: str | None = None
    ats_detection_error: str | None = None
    ats_difficulty: ATSDifficultyEnum | None = None


class ApplySingleRequest(BaseModel):
    application_id: uuid.UUID


class ApplySingleResponse(BaseModel):
    application_id: uuid.UUID
    task_id: str
    message: str


class ApplyBulkRequest(BaseModel):
    application_ids: list[uuid.UUID] = Field(..., max_length=10)


class ApplyBulkResponse(BaseModel):
    bulk_task_id: str
    application_ids: list[uuid.UUID]
    message: str


class ApplyStatusResponse(BaseModel):
    application_id: uuid.UUID
    status: str
    step: str | None = None
    error: str | None = None
    notes: str | None = None
    resume_id: uuid.UUID | None = None
    cover_letter_id: uuid.UUID | None = None
    screenshot_path: str | None = None
    manual_url: str | None = None


class ApplyBulkStatusResponse(BaseModel):
    bulk_task_id: str
    total: int
    completed: int
    failed: int
    manual_required: int
    pending: int
    results: list[ApplyStatusResponse]


class ApplyOrchestratorResult(BaseModel):
    application_id: uuid.UUID
    success: bool
    status: str
    resume_id: uuid.UUID | None = None
    cover_letter_id: uuid.UUID | None = None
    screenshot_path: str | None = None
    manual_url: str | None = None
    error: str | None = None
    steps_completed: list[str]
    total_latency_ms: int


class ApplySSEEvent(BaseModel):
    event: str
    application_id: uuid.UUID
    step: str | None = None
    status: str | None = None
    error: str | None = None
    notes: str | None = None
    manual_url: str | None = None

File: frontend/src/types/apply.ts
export type ApplyStep =
  | "load_profile"
  | "generate_resume"
  | "save_resume"
  | "generate_cover_letter"
  | "filling_form"
  | "submitting"
  | "verifying";

export const APPLY_STEP_LABELS: Record<ApplyStep, string> = {
  load_profile: "Loading profile",
  generate_resume: "Generating resume",
  save_resume: "Saving resume",
  generate_cover_letter: "Generating cover letter",
  filling_form: "Filling application form",
  submitting: "Submitting application",
  verifying: "Verifying submission",
};

export const APPLY_STEPS: ApplyStep[] = [
  "load_profile",
  "generate_resume",
  "save_resume",
  "generate_cover_letter",
  "filling_form",
  "submitting",
  "verifying",
];

export type ApplySinglePhase = "detecting" | "applying" | "manual_required";

export interface ApplySingleRequest {
  application_id: string;
}

export interface ApplySingleResponse {
  application_id: string;
  task_id: string;
  message: string;
  phase?: ApplySinglePhase | null;
}

export interface ATSDetectionStatusResponse {
  application_id: string;
  job_id: string;
  status: string;
  notes: string | null;
  ats_platform: string | null;
  ats_detection_method: string | null;
  ats_confidence: number | null;
  ats_form_url: string | null;
  ats_detected_fields: string[] | null;
  ats_screenshot_path: string | null;
  ats_detection_error: string | null;
  ats_difficulty: string | null;
}

export interface ApplyBulkRequest {
  application_ids: string[];
}

export interface ApplyBulkResponse {
  bulk_task_id: string;
  application_ids: string[];
  message: string;
}

export interface ApplyStatusResponse {
  application_id: string;
  status: string;
  step: string | null;
  steps_completed: string[] | null;
  error: string | null;
  notes: string | null;
  resume_id: string | null;
  cover_letter_id: string | null;
  screenshot_path: string | null;
  manual_url: string | null;
}

export interface ApplyBulkStatusResponse {
  bulk_task_id: string;
  total: number;
  completed: number;
  failed: number;
  manual_required: number;
  pending: number;
  results: ApplyStatusResponse[];
}

export interface ApplySSEEvent {
  event: string;
  application_id: string;
  step: string | null;
  steps_completed?: string[] | null;
  status: string | null;
  error: string | null;
  notes?: string | null;
  manual_url?: string | null;
}
